### PR TITLE
Scaffold

### PR DESCRIPTION
**Pull request by bwighane**

### CODE REVIEW

File: src/helpers/model-helpers.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { Sequelize } from "sequelize";
```
Comment (liwoo) [inline src/helpers/model-helpers.ts]: validation-worker
  => modules
       -> validation.ts
       -> migration.ts
       -> queue.ts

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  connection.createChannel(function (error: Error, channel: Channel) {
+    var queueName = process.env.AVW_QUEUE_NAME || "DHIS2_VALIDATION_QUEUE";
+    channel.assertQueue(queueName, { durable: true });
+    console.log("[*] Waiting for messages on %s. To exit press CTRL+C", queueName);
+    channel.consume(queueName, async function (message: Message | null) {
+      if (message) {
+        const queueMessage: QueueMessage = JSON.parse(message.content.toString());
+        const logger = new Logger(queueMessage.channelId);
+        const sequelize = await connectToDatabase(
+          process.env.AVW_DATABASE_HOST || '',
```
Comment (liwoo) [inline src/index.ts]: use defaults like root:password

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        const migration = await recordStartMigration(sequelize, clientId);
+
+        const payloadFile = `${process.env.AVW_PAYLOADS_ROOT_DIR}/${queueMessage.channelId}.adx`;
+        const payload = JSON.parse(readFileSync(payloadFile).toString());
+        const { error } = Joi.validate(payload, PayloadSchema);
+
+        if (!error) {
+          await recordStructureValidationStatus(sequelize, migration.get('id'), true);
+          logger.info('Payload passed structure validation validation');
+          //TODO: start off on the other validation - authorization
```
Comment (liwoo) [inline src/index.ts]: are these todos being done?

**Review by liwoo — CHANGES_REQUESTED** [review]

**Review by liwoo — COMMENTED** [review]
Lets create an ADXLogger which will be listening to RabitMQ and we will simply call it from all the workers (and mediator) that need to log stuff

File: src/helpers/model-helpers.ts (hunk 1 above)
Comment (bwighane) [reply src/helpers/model-helpers.ts]: This is duly noted, I propose that perhaps these should come as a refactor after merging with my current working branch.

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+          process.env.AVW_DATABASE || '',
+          process.env.AVW_DATABASE_USERNAME || 'username',
+          process.env.AVW_DATABASE_PASSWORD || ''
+        );
+        const clientId = await getClientId(sequelize, queueMessage.clientId);
+
+        const migration = await recordStartMigration(sequelize, clientId);
+
+        const payloadFile = `${process.env.AVW_PAYLOADS_ROOT_DIR}/${queueMessage.channelId}.adx`;
+        const payload = JSON.parse(readFileSync(payloadFile).toString());
```
Comment (liwoo) [inline src/index.ts]: what happens if we can't open the file, or it doesn't exist? we should probably handle this

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        const { error } = Joi.validate(payload, PayloadSchema);
+
+        if (!error) {
+          await recordStructureValidationStatus(sequelize, migration.get('id'), true);
+          logger.info('Payload passed structure validation validation');
+          //TODO: start off on the other validation - authorization
+        } else {
```
Comment (liwoo) [inline src/index.ts]: no need for this else statement here...

File: src/models/Facility.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  openLMISFacilityCode: sequelize.STRING
+}
+
+const options = {
+  freezeTableName: true,
+  tableName,
+  timestamps: false
+}
+
+export const createProductModel = async (
```
Comment (liwoo) [inline src/models/Facility.ts]: should this be createFacilityModel?

File: src/modules/models.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { Sequelize } from "sequelize";
```
Comment (liwoo) [inline src/modules/models.ts]: lets call this file databaseOperations

**Review by liwoo — CHANGES_REQUESTED** [review]

File: src/models/Facility.ts (hunk 6 above)
Comment (bwighane) [reply src/models/Facility.ts]: It Should, I was not using it at this point so I realised at a later stage in my current working branch

File: src/index.ts (hunk 4 above)
Comment (bwighane) [reply src/index.ts]: Handled now

**Review by bwighane — COMMENTED** [review]
I've made the necessary changes. With regards to line 44 in the main index.ts file, I need to add some logic that only executes when structure validation fails. Please advise further.

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+import {
+  recordStartMigration,
+  getClientId,
+  recordStructureValidationStatus,
+}
+  from './modules';
+
+amqp.connect(process.env.AVW_QUEUE_HOST || 'amqp://localhost', function (error: Error, connection: Connection) {
+  connection.createChannel(function (error: Error, channel: Channel) {
+    var queueName = process.env.AVW_QUEUE_NAME || "DHIS2_VALIDATION_QUEUE";
```
Comment (liwoo) [inline src/index.ts]: this should be a `const` please and define it outside amqp.connect

File: src/index.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+import * as Joi from 'joi';
+const utils = require('utils')._;
+import {
+  recordStartMigration,
+  getClientId,
+  recordStructureValidationStatus,
+}
+  from './modules';
+
+amqp.connect(process.env.AVW_QUEUE_HOST || 'amqp://localhost', function (error: Error, connection: Connection) {
```
Comment (liwoo) [inline src/index.ts]: this connection string should be a variable in a seperate line

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  connection.createChannel(function (error: Error, channel: Channel) {
+    var queueName = process.env.AVW_QUEUE_NAME || "DHIS2_VALIDATION_QUEUE";
+    channel.assertQueue(queueName, { durable: true });
+    console.log("[*] Waiting for messages on %s. To exit press CTRL+C", queueName);
+    channel.consume(queueName, async function (message: Message | null) {
+      if (message) {
```
Comment (liwoo) [inline src/index.ts]: start with the negative null case and end with the happy path (again, no need for an else statement)

**Review by liwoo — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  getClientId,
+  recordStructureValidationStatus,
+}
+  from './modules';
+
+const main = () => {
+  const host = process.env.AVW_QUEUE_HOST || 'amqp://localhost';
+  amqp.connect(host, function (error: Error, connection: Connection) {
+    connection.createChannel(function (error: Error, channel: Channel) {
+      var queueName = process.env.AVW_QUEUE_NAME || "DHIS2_VALIDATION_QUEUE";
```
Comment (liwoo) [inline src/index.ts]: this var should be a const

File: src/index.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+      channel.assertQueue(queueName, { durable: true });
+      console.log("[*] Waiting for messages on %s. To exit press CTRL+C", queueName);
+      channel.consume(queueName, async function (message: Message | null) {
+        if (!message) {
+          return
+        }
+        const queueMessage: QueueMessage = JSON.parse(message.content.toString());
+        const logger = new Logger(queueMessage.channelId);
+        const sequelize = await connectToDatabase(
```
Comment (liwoo) [inline src/index.ts]: connect to db outside the `amqp.connect`..

File: src/index.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
+import * as  amqp from 'amqplib/callback_api';
+import { Channel, Connection, Message } from 'amqplib/callback_api';
+import { config } from 'dotenv'; config(); //initialize right there the configuration
+import { QueueMessage } from './types';
+import { PayloadSchema } from './schemas';
+import { connectToDatabase } from './datasource';
+import { Logger } from './utils';
+import * as Joi from 'joi';
+const utils = require('utils')._;
```
Comment (liwoo) [inline src/index.ts:9]: what's this doing?

**Review by liwoo — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 13 above)
Comment (bwighane) [reply src/index.ts:9]: `const payloadFileContent = utils.tryRead(payloadFile);`

That's where the utils module is being used. It uses the old type of node exports hence cannot be used with the current Typescript way of importing. The logic is encapsulated inside the _ property of the module.

Comment (bwighane) [reply src/index.ts:9]: It's a safer/shorter way of handling file reading errors.

Comment (liwoo) [reply src/index.ts:9]: :ok:

**Review by liwoo — APPROVED** [review]

**Pull request merged**